config/src/runtime_upgrades.rs

```diff
@@ -0,0 +1,278 @@
+{
+    fn from(
+        _: PoisonError<
+            MutexGuard<'_, std::option::Option<Box<(dyn Reloadable<T> + Send + Sync + 'static)>>>,
+        >,
+    ) -> Self {
+        Self::PoisonedMutex
+    }
+}
+
+/// Implement this trait if the field must be mutably borrowed for `reload`.
+pub trait Reloadable<T> {
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of
+    /// [`ReloadFailure`]. [`ReloadFailure::Other`] can be used as a
+    /// **temporary** placeholder.
+    fn reload(&mut self, item: T) -> Result<()>;
+}
+
+/// Implement this trait if the object can be immutably borrowed and `reload`-ed.
+pub trait InnerMutablyReloadable<T> {
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of [`ReloadFailure`].
+    /// [`ReloadFailure::Other`] can be used as a **temporary** placeholder.
+    fn reload(&self, item: T) -> Result<()>;
+}
+
+impl<T> std::fmt::Debug for dyn Reloadable<T> {
+    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl<T> std::fmt::Debug for dyn Reloadable<T> + Send + Sync {
+    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl<T> Reloadable<T> for dyn InnerMutablyReloadable<T> {
+    fn reload(&mut self, item: T) -> Result<(), ReloadFailure> {
+        InnerMutablyReloadable::reload(self, item)
+    }
+}
+
+/// Contains [`handle`] types: opaque wrappers around a reloadable
+/// configuration.  [`handle::Reconfigurable`] is the most commonly
+/// useful type of handle, as it deserializes into the inner object,
+/// as well as handles synchronisation of changes to the program state
+/// via the handle, to the changes in the configuration value.
+pub mod handle {
+    use std::{
+        fmt::Debug,
+        sync::{Arc, Mutex},
+    };
+
+    use crossbeam::atomic::AtomicCell;
+    use serde::{Deserialize, Serialize};
+
+    use super::{InnerMutablyReloadable, ReloadFailure, Reloadable, Result};
+    // -----------------------------------------------------------------
+
+    /// When an object with runtime upgradeable configuration is created,
+    /// sometimes it is necessary to store some transient state in order
+    /// to later be able to quickly and efficiently reload the
+    /// configuration. This handle must implement [`Reloadable<T>`] and be
+    /// stored in [`SideEffect<T>`], using [`ReloadHandle::set`].
+    ///
+    /// If the Handle is stored with the configuration, each cloned
+    /// instance contains an [`std::sync::Arc`] with a [`std::sync::Mutex`], so the
+    /// containing object (configuration) doesn't need to be `mut`.
+    ///
+    /// [`SideEffect`] uses dynamic dispatch (intentionally) in cases
+    /// (e.g. [`tracing_subscriber`]) where the reloading process makes
+    /// use of impractically many generic arguments. If this is not the
```

Comment:
did you consider using enum instead of trait object? They are interchangeable when you know ahead of time all possible implementors of your trait.

I have no idea if this is applicable here, just pointing this out